Add `inheritFrom` and `css` to JSON editor

rogerl50 opened the pull request:
Done and ready for review.

Original note:


`inheritFrom` I believe to be complete and working.

For `css`, I'm looking for feedback as to whether this works as expected. Currently there is only really support for the widget-specific stuff for `timer` and `seat`. What I want feedback on is:
1. Anything that doesn't work at all, and comments on the code.
2. Is this the kind of interaction y'all are looking for?
3. Should there be an "add field" at each level to just put in an empty css class or property, or should we just leave that to be typed manually?

Note that conversion from the string form to the object form is supported for both `inheritFrom` and `css`.

ArnoldSmith86 commented:
PR-SERVER-BOT: You can play around with it here: https://test.virtualtabletop.io/PR-1335/pr-test (or any other room on that server)

After merging, a backup will be available at https://beta.virtualtabletop.io/editor/PR1335-pr-test.

rogerl50 commented:
I have a handle on how to do whatever needs to be done to properly support css in the editor. But it's not clear to me what the scope of the support should be. For example, `canvas` and `holder` currently have the following default css:
```JSON
.canvas {
  border-width: calc(1px / var(--scale));
  border-style: solid;
  border-color: var(--VTTblue);
}

.holder {
  background: white;
  box-sizing: border-box;
  border: calc(1px / var(--scale)) solid #333;
  border-color: #d8d8d8 #ccc #ccc #d8d8d8;
}
```
Do we want these to appear in the defaults for these two widget classes? How about
```JSON
.deck {
  display: none;
}
```
Holder gets more complicated:
```JSON
.holder {
  background: white;
  box-sizing: border-box;
  border: calc(1px / var(--scale)) solid #333;
  border-color: #d8d8d8 #ccc #ccc #d8d8d8;
}

.holder > .widget, .widget.droppable.droptarget.holder.transparent > .widget {
  left: -1px;
  top: -1px;
}

.holder.transparent {
  background: none;
  border: none;
}

.holder.transparent > .widget {
  left: 0;
  top: 0;
}

.widget.droppable.droptarget {
  border: calc(1px / var(--scale)) solid #333 !important;
}
```

Obviously (I guess) we aren't going to support all this generality in the editor. But it'd be nice to have some guidelines overall for what we *do* want to support. Alternatively, I can just code up something that provides at least the current functionality, but in the object form, we can merge that and improve over time. Suggestions?

ArnoldSmith86 commented:
Personally, I wouldn't add anything other than what you already have: color, background, font-size especially.

All those classes we talked about like `droppable` should be there as commands just so you don't have to remember them but I wouldn't add different CSS properties for those. Just add them as an empty object.

What I'm still missing is the simple object format:

```JSON
"css": {
  "color": "white"
}
```

I feel like that is the format that should be used ~90% of the time and as far as I can tell there are currently no commands for that.

96LawDawg commented:
Some feedback:

- I agree with with @ArnoldSmith86 says above.
- When adding css, if you include both "hover" and "default," the blue buttons get duplicated (one for each of hover and default).  There should be only 1; the one for the sub-property that you are in.
- Default font-size should not be 30px.  It should be whatever the actual default is.  16px I think?
- For `inheritFrom` it seems odd to include height and width as the defaults.  While it is true that they may be the most common, if you are manually adding `inheritFrom` on a widget, it already has height and width.  But to get that to work, you need to manually delete the existing height and width.  I would just put blank quotation marks where you have the default properties.  We can type the properties we want.

rogerl50 commented:
The latest update makes the simple object form
```JSON
"css": {
  "color": "white"
}
```
the default.

It also removes the duplicated options from the `default` and `:hover` items, and fixed the default `font-size` (although I will note that `30px` is the current default if you add `font-size` to `css`).

Remains to deal with widget-specific items.

96LawDawg commented:
From #1337...

> > Also, for css, what about adding a "background-image" button? I use that quite a bit and always have to go and cut and paste it from somewhere else. But maybe that is just me. If it is included, it would be most useful if it also included several other related settings. To be clear, I'm only asking for the 1 button, but pressing the button gives you this:
> > `"background-image": "url(' ')", "background-size": "contain", "background-repeat": "no-repeat", "background-position": "center"`
> > I assume you mean for buttons? Or for everything? Should it be in `default`, `:hover`, or somewhere else? Also, are those other properties (from `background-size` on) used in other contexts than with images?
> 

One button shows up when the context is in `default` or when it is a simple object form.  The button is labeled `background-image`.  When you press it, you get those 4 default settings listed above.  According to https://www.w3schools.com/cssref/default.asp, those other 3 properties are only used in connection with `background-image`. That is the only time I've seen them used as well, but I'm not a CSS expert.

rogerl50 commented:
> One button shows up when the context is in `default` or when it is a simple object form. The button is labeled `background-image`. When you press it, you get those 4 default settings listed above. According to https://www.w3schools.com/cssref/default.asp, those other 3 properties are only used in connection with `background-image`. That is the only time I've seen them used as well, but I'm not a CSS expert.

That seems complicated. So should I show that button only when none of the four is already in the css? Only when they all aren't there? This is also the only case so far where a button would produce more than one property in the result.

I'm inclined to leave this to the game designer.

96LawDawg commented:
> I'm inclined to leave this to the game designer.

Then how about one button just for background-image that produces the background-image property in the format above?

ArnoldSmith86 commented:
BTW, please don't change widget property defaults _unless absolutely necessary_. I'm pretty sure that's the reason why TestCafé fails.

rogerl50 commented:
> BTW, please don't change widget property defaults _unless absolutely necessary_. I'm pretty sure that's the reason why TestCafé fails.

I changed the `css` and `inheritFrom` property defaults in widget.js to `{}`. I also moved the various color names (like `backgroundColorOH`) from the constructor into instance variables so they don't show up in the context commands any more. That all seemed necessary to me; do you agree?

robartsd commented:
> I changed the `css` and `inheritFrom` property defaults in widget.js to `{}`. I also moved the various color names (like `backgroundColorOH`) from the constructor into instance variables so they don't show up in the context commands any more. That all seemed necessary to me; do you agree?

It is hard to imagine a case where changing these properties from defaulting to empty strings to defaulting to empty objects would break a game; however, there is the remote possibility that game automation gets one of these properties and will get tripped up by the form being different than expected. Of course, such could be resolved with a file updater that explicitly sets the properties to the old defaults for every widget that does not already have the property; but that would be less convenient than the status quo for game designers working on games that got updated.

ArnoldSmith86 commented:
Any button that appends CSS strings to `css`, where it might be unset previously, would probably break now. I'm guessing that's why TestCafé fails.

Changing a property default is always risky IMO and not worth it just for keeping the editor code a bit simpler.

I hope, I don't sound rude. Your approach is of course cleaner but I don't think it's a good idea.

About the changes to `button.js`: removing the property defaults doesn't technically change anything because `null` is the default default but I would like to have all properties in that defaults object that are used by the engine. Your new `let ...` lines don't do anything either because they are not used anywhere and actually are only defined in the constructor and nowhere else. _Maybe_ just comment out the lines in that object? But honestly, I would prefer to exclude them in the editor and keep them in the defaults.

rogerl50 commented:
> Any button that appends CSS strings to `css`, where it might be unset previously, would probably break now. I'm guessing that's why TestCafé fails.
> 
> Changing a property default is always risky IMO and not worth it just for keeping the editor code a bit simpler.
> 
> I hope, I don't sound rude. Your approach is of course cleaner but I don't think it's a good idea.
> 
> About the changes to `button.js`: removing the property defaults doesn't technically change anything because `null` is the default default but I would like to have all properties in that defaults object that are used by the engine. Your new `let ...` lines don't do anything either because they are not used anywhere and actually are only defined in the constructor and nowhere else. _Maybe_ just comment out the lines in that object? But honestly, I would prefer to exclude them in the editor and keep them in the defaults.

I assume we still want the editor defaults to show up as `{}` rather than strings (that will require some special case code). Making the changes for the button object will also require special case code. I will go ahead and do that.

ArnoldSmith86 commented:
I played around with it a bit. Very nice stuff.

I'll have a closer look at the code changes when you're "done" and the PR is "Ready for review". Any features you still want to add here?

rogerl50 commented:
I think this is finally ready for review. The structure is (I hope) such that we can easily add additional css sections for particular widget types if desired.

96LawDawg reviewed (approved):
Approved through test.

ArnoldSmith86 commented on client/js/jsonedit.js:
```diff
@@ -830,35 +897,89 @@ function displayComputeOps() {
+          jeSetAndSelect({});
         }
       });
+      for(const cssProperty of Object.keys(nested_presets[type][cssSection])) { // Add entries per-section
```
Same question about `for(... in ...)`.

And since you use the key _and_ the value in your code, you could also consider using `for(const [ key, value ] of Object.entries(object))`.

ArnoldSmith86 commented on client/js/jsonedit.js:
```diff
@@ -830,35 +897,89 @@ function displayComputeOps() {
   }
+
+  // Add simple object button items
+  for(const cssProperty of Object.keys(string_presets)) { // Add entries in "default" (only) section
```
`of` instead of `in` again.

ArnoldSmith86 commented on client/js/jsonedit.js:
```diff
@@ -830,35 +897,89 @@ function displayComputeOps() {
+      name: cssProperty,
+      context: `.* ↦ (css|[a-z]+CSS)`,
+      show: function() { // Need to make sure it is a simple object (contents are "key": "string" pairs)
+        const cssKind = jeContext.join(' ↦ ').match(this.context)[1];
```
This should throw an exception every time the context does _not_ match. It should be caught and everything should be fine. I don't think I mind.

rogerl50 replied:
But it should always match, since that's how we got here.

ArnoldSmith86 commented on client/js/jsonedit.js:
```diff
@@ -1937,50 +2065,50 @@ function jeSetEditorContent(content) {
       const name = (typeof command.name == 'function' ? command.name() : command.name);
       const icon = (typeof command.icon == 'function' ? command.icon() : command.icon);
       const isMaterial = String(icon).match(/^[^[]/) ? 'material' : '';
       commandText += `<button class='top ${isMaterial}' id='${command.id}' title='${name}' ${!command.show || command.show() ? '' : 'disabled'}>${icon}</button>`;
     }
   }
   commandText += `</div>`;
-  delete activeCommands[""];
+
+  // Next figure out which context commands are active here.
+  const activeCommands = {};
+  const context = jeContext.join(' ↦ ');
+  for(const command of jeCommands) {
+    delete command.currentKey;
+    const contextMatch = context.match(new RegExp(command.context));
+    if(contextMatch && contextMatch[0]!= "" && (!command.context || jeStateNow && !jeJSONerror) && (!command.show || command.show())) {
+      if(activeCommands[contextMatch[0]] === undefined)
+        activeCommands[contextMatch[0]] = [];
+      activeCommands[contextMatch[0]].push(command);
+    };
+  }
 
   if(jeContext[jeContext.length-1] == '(var expression)') {
     commandText += `\n  <b>var expression</b>\n<label>Search </label><input id="var_search" name="var_search" type="text"><br>`;
     commandText += `<div id="var_results"></div>\n`;
   }
 
   if(!jeJSONerror && jeStateNow) {
+    const usedKeys = { a: 1, c: 1, x: 1, v: 1, w: 1, n: 1, t: 1, q: 1, j: 1, z: 1 };
+
+    const sortByName = function(a, b) {
+      const nameA = typeof a.name == 'function' ? a.name() : a.name;
+      const nameB = typeof b.name == 'function' ? b.name() : b.name;
+      return nameA.localeCompare(nameB);
+    }
+
+    const displayKey = function (k) {
+      return { ArrowUp: '⬆', ArrowDown: '⬇'} [k] || k;
+    }
+
```
As far as I can tell, you pretty much only rearranged this, right? This kind of stuff is always so hard to decipher in the diff...

rogerl50 replied:
That's correct. Assuming I did it properly :)

ArnoldSmith86 reviewed (approved):
Code looks fine to me. I left a few comments but those are more FYI than change requests.

rogerl50 commented on client/js/jsonedit.js:
```diff
@@ -232,6 +232,73 @@ const jeCommands = [
+    name: 'convert to simple object',
+    context: '^.* ↦ (css|[a-z]+CSS)',
+    show: function() {
+      const cssKind = jeContext.join(' ↦ ').match(this.context)[1];
+      return typeof jeStateNow[cssKind] == "string";
```
As noted below, it is supported separately.

rogerl50 commented on client/js/jsonedit.js:
```diff
@@ -232,6 +232,73 @@ const jeCommands = [
+      return true;
+    },
+    call: async function() {
+      const newCSS = {};
+      for(const property in jeStateNow.css)
+        newCSS[property] = jeStateNow.css[property];
```
Yeah. Simpler.

rogerl50 commented on client/js/jsonedit.js:
```diff
@@ -830,35 +897,89 @@ function displayComputeOps() {
+
+  // Add nested object button items
+  for(const type in nested_presets) {
+    for(const cssSection of Object.keys(nested_presets[type])) { // Add CSS sections
```
Done.
